// @dainprotocol/utils/src/generateUI/base/base.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DSLBuilder = void 0;
class DSLBuilder {
    constructor(type, initialProps, renderMode = "inline") {
        this.element = {
            type,
            props: Object.assign({}, initialProps),
            children: [],
            renderMode,
            actions: [],
        };
    }
    setProp(key, value) {
        this.element.props[key] = value;
        return this;
    }
    setRenderMode(mode) {
        this.element.renderMode = mode;
        return this;
    }
    addChild(child) {
        if (child instanceof DSLBuilder) {
            this.element.children.push(child.build());
        }
        else {
            this.element.children.push(child);
        }
        return this;
    }
    // NEW: Add an action to this element
    addAction(action) {
        if (!this.element.actions) {
            this.element.actions = [];
        }
        this.element.actions.push(action);
        return this;
    }
    build() {
        return this.element;
    }
}
exports.DSLBuilder = DSLBuilder;
