// @dainprotocol/utils/src/generateUI/builders/server/imageGallery/imageGallery.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageGalleryUIBuilder = void 0;
const base_builder_1 = require("../../../base/base.builder");
class ImageGalleryUIBuilder extends base_builder_1.DSLBuilder {
    constructor(initialProps, renderMode = "inline") {
        super("imageGallery", Object.assign({ images: [] }, initialProps), renderMode);
    }
    /**
     * Sets the title for the gallery
     */
    title(title) {
        this.setProp("title", title);
        return this;
    }
    /**
     * Sets the description for the gallery
     */
    description(description) {
        this.setProp("description", description);
        return this;
    }
    /**
     * Sets the number of columns for the gallery layout
     */
    columns(columns) {
        this.setProp("columns", columns);
        return this;
    }
    /**
     * Adds a single image to the gallery
     */
    addImage(image) {
        this.element.props.images.push(image);
        return this;
    }
    /**
     * Adds multiple images at once
     */
    addImages(images) {
        images.forEach(image => this.addImage(image));
        return this;
    }
    /**
     * Defines the action to be triggered when an image is clicked
     */
    onViewImage(action) {
        const { tool, params, paramSchema } = action;
        if (!tool) {
            throw new Error("onViewImage action must specify a 'tool'");
        }
        return this.addAction({
            type: "callTool",
            tool,
            params,
            paramSchema,
            key: "view"
        });
    }
}
exports.ImageGalleryUIBuilder = ImageGalleryUIBuilder;
