// @dainprotocol/utils/src/generateUI/builders/server/card/card.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CardUIBuilder = void 0;
const base_builder_1 = require("../../../base/base.builder");
class CardUIBuilder extends base_builder_1.DSLBuilder {
    constructor(initialProps, renderMode = "inline") {
        super("card", initialProps, renderMode);
    }
    title(title) {
        return this.setProp("title", title);
    }
    content(content) {
        return this.setProp("content", content);
    }
    onConfirm(action) {
        const { tool, paramSchema, params } = action;
        if (!tool || !paramSchema || !params) {
            throw new Error("onConfirm action must have a tool, paramSchema, and params");
        }
        return this.addAction({
            type: "callTool",
            tool,
            paramSchema,
            params,
            key: "confirm",
        });
    }
    addField(field) {
        this.element.props.fields = this.element.props.fields || [];
        this.element.props.fields.push(field);
        return this;
    }
    addFields(fieldsOrItems, mapper) {
        if (mapper) {
            fieldsOrItems.forEach((item) => {
                const field = mapper(item);
                this.addField(field);
            });
        }
        else {
            fieldsOrItems.forEach((field) => {
                this.addField(field);
            });
        }
        return this;
    }
}
exports.CardUIBuilder = CardUIBuilder;
