// @dainprotocol/utils/src/generateUI/builders/server/payment/payment.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentUIBuilder = void 0;
const base_builder_1 = require("../../../base/base.builder");
class PaymentUIBuilder extends base_builder_1.DSLBuilder {
    constructor(paymentSuccess, initialProps, renderMode = "inline") {
        super("payment", Object.assign({ paymentSuccess }, initialProps), renderMode);
    }
    /**
     * Sets the status message
     */
    setMessage(message) {
        this.setProp("message", message);
        return this;
    }
    /**
     * Sets the transaction details
     */
    setTransactionDetails(transactionId, amount, currency) {
        this.setProp("transactionId", transactionId);
        this.setProp("amount", amount);
        this.setProp("currency", currency);
        return this;
    }
    /**
     * Defines the action to be triggered when retrying a failed payment
     */
    onRetry(action) {
        const { tool, params, paramSchema } = action;
        if (!tool) {
            throw new Error("onRetry action must specify a 'tool'");
        }
        return this.addAction({
            type: "callTool",
            tool,
            params,
            paramSchema,
            key: "retry"
        });
    }
    /**
     * Defines the action to be triggered when initiating a payment
     */
    onPay(action) {
        const { tool, params, paramSchema } = action;
        if (!tool) {
            throw new Error("onPay action must specify a 'tool'");
        }
        return this.addAction({
            type: "callTool",
            tool,
            params,
            paramSchema,
            key: "pay"
        });
    }
}
exports.PaymentUIBuilder = PaymentUIBuilder;
