// @dainprotocol/utils/src/generateUI/builders/server/chart/chart.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChartUIBuilder = void 0;
const base_builder_1 = require("../../../base/base.builder");
class ChartUIBuilder extends base_builder_1.DSLBuilder {
    constructor(initialProps, renderMode = "inline") {
        super("chart", initialProps, renderMode);
    }
    /**
     * Sets the chart type (bar, line, or pie).
     */
    type(type) {
        return this.setProp("type", type);
    }
    /**
     * Sets the title of the chart.
     */
    title(title) {
        return this.setProp("title", title);
    }
    /**
     * Sets an optional description for the chart.
     */
    description(description) {
        return this.setProp("description", description);
    }
    /**
     * Sets the data array for the chart.
     */
    chartData(data) {
        return this.setProp("data", data);
    }
    /**
     * Sets the data keys for mapping data to chart axes.
     */
    dataKeys(dataKeys) {
        return this.setProp("dataKeys", dataKeys);
    }
    /**
     * Sets optional trend information.
     */
    trend(value, text) {
        return this.setProp("trend", { value, text });
    }
    /**
     * Sets an optional footer text.
     */
    footer(footer) {
        return this.setProp("footer", footer);
    }
}
exports.ChartUIBuilder = ChartUIBuilder;
