// @dainprotocol/utils/src/generateUI/builders/server/cardList/cardList.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CardListUIBuilder = void 0;
const base_builder_1 = require("../../../base/base.builder");
/**
 * DSL builder for creating a "cardList" element.
 */
class CardListUIBuilder extends base_builder_1.DSLBuilder {
    constructor(initialProps, renderMode = "inline") {
        super("cardList", Object.assign({ cards: [] }, initialProps), renderMode);
    }
    /**
     * Sets the top-level title for the card list.
     */
    title(title) {
        this.setProp("title", title);
        return this;
    }
    /**
     * Sets a description for the card list.
     */
    description(desc) {
        this.setProp("description", desc);
        return this;
    }
    /**
     * Add a single card to the list.
     */
    addCard(card) {
        this.element.props.cards.push(card);
        return this;
    }
    /**
     * Add multiple cards at once.
     */
    addCards(cards) {
        cards.forEach((card) => this.addCard(card));
        return this;
    }
    /**
     * Defines the tool/action to be triggered on a user confirmation from the card list.
     */
    onConfirm(action) {
        const { tool, params } = action;
        if (!tool) {
            throw new Error("onConfirm action must specify a 'tool' (callTool).");
        }
        this.addAction({
            type: "callTool",
            tool,
            params,
            key: "confirm",
        });
        return this;
    }
    /**
     * Provide DSL children to this card list element.
     */
    withChildren(child) {
        this.addChild(child);
        return this;
    }
}
exports.CardListUIBuilder = CardListUIBuilder;
