// @dainprotocol/utils/src/generateUI/builders/server/alert/alert.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlertUIBuilder = void 0;
const base_builder_1 = require("../../../base/base.builder");
class AlertUIBuilder extends base_builder_1.DSLBuilder {
    constructor(initialProps, renderMode = "inline") {
        super("alert", initialProps, renderMode);
    }
    /**
     * Sets the variant for the alert.
     * @param variant The alert variant (info, warning, error, or success)
     */
    variant(variant) {
        return this.setProp("variant", variant);
    }
    /**
     * Sets a custom message for the alert.
     */
    message(message) {
        return this.setProp("message", message);
    }
    /**
     * Sets the title for the alert.
     */
    title(title) {
        return this.setProp("title", title);
    }
    /**
     * Sets whether to show an icon.
     */
    icon(show = true) {
        return this.setProp("icon", show);
    }
    /**
     * Sets the alert type.
     */
    type(type) {
        return this.setProp("type", type);
    }
}
exports.AlertUIBuilder = AlertUIBuilder;
