// @dainprotocol/utils/src/generateUI/builders/server/progressList/progressList.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressListUIBuilder = void 0;
const base_builder_1 = require("../../../base/base.builder");
class ProgressListUIBuilder extends base_builder_1.DSLBuilder {
    constructor(initialProps, renderMode = "inline") {
        super("progressList", Object.assign({ items: [] }, initialProps), renderMode);
    }
    /**
     * Sets the title for the progress list
     */
    title(title) {
        this.setProp("title", title);
        return this;
    }
    /**
     * Adds a single progress item to the list
     */
    addItem(item) {
        this.element.props.items.push(item);
        return this;
    }
    /**
     * Adds multiple progress items at once
     */
    addItems(items) {
        items.forEach(item => this.addItem(item));
        return this;
    }
    /**
     * Defines the action to be triggered when an item is selected
     */
    onSelectItem(action) {
        const { tool, params, paramSchema } = action;
        if (!tool) {
            throw new Error("onSelectItem action must specify a 'tool'");
        }
        return this.addAction({
            type: "callTool",
            tool,
            params,
            paramSchema,
            key: "select"
        });
    }
}
exports.ProgressListUIBuilder = ProgressListUIBuilder;
