// @dainprotocol/utils/src/DainResponse/DainResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DainResponse = void 0;
const generateUI_1 = require("../generateUI");
class DainResponse {
    constructor({ data, text, ui, pleasePay, processes }) {
        this.data = {};
        this.text = "";
        this.data = data;
        this.text = text !== null && text !== void 0 ? text : "";
        // If ui is a DSLBuilder or an array of them, build accordingly
        if (Array.isArray(ui)) {
            this.ui = ui.map((item) => item instanceof generateUI_1.DSLBuilder ? item.build() : item);
        }
        else {
            this.ui = ui instanceof generateUI_1.DSLBuilder ? ui.build() : ui;
        }
        this.pleasePay = pleasePay;
        this.processes = processes;
    }
    /**
     * Static constructor to help build a DainResponse in a more fluent style.
     */
    static create(options) {
        return new DainResponse(options);
    }
    /**
     * Example method for returning JSON, similar to NextResponse.json().
     */
    toJSON() {
        return {
            data: this.data,
            text: this.text,
            ui: this.ui,
            pleasePay: this.pleasePay,
            processes: this.processes,
        };
    }
}
exports.DainResponse = DainResponse;
