// @dainprotocol/utils/src/generateUI/builders/server/orderStatus/orderStatus.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderStatusUIBuilder = void 0;
const base_builder_1 = require("../../../base/base.builder");
/**
 * DSL builder for creating an "orderStatus" element.
 */
class OrderStatusUIBuilder extends base_builder_1.DSLBuilder {
    constructor(initialProps, renderMode = "inline") {
        super("orderStatus", Object.assign({}, initialProps), renderMode);
    }
    /**
     * Define or update the status string (e.g., "Pending", "Out for Delivery").
     */
    setStatus(status) {
        this.setProp("status", status);
        return this;
    }
    /**
     * Set or update the address where the order is being delivered.
     */
    setAddress(address) {
        this.setProp("address", address);
        return this;
    }
    /**
     * Modify the final quote details, including subtotal, fees, etc.
     */
    setFinalQuote(finalQuote) {
        this.setProp("finalQuote", finalQuote);
        return this;
    }
    /**
     * Provide or update tracking information, e.g., a link.
     */
    setTracking(tracking) {
        this.setProp("tracking", tracking);
        return this;
    }
    /**
     * Add or modify driver info, e.g., driver's name or phone number.
     */
    setDriver(driver) {
        this.setProp("driver", driver);
        return this;
    }
    /**
     * Update store information, e.g., store name and its logo image.
     */
    setStoreInfo(storeInfo) {
        this.setProp("storeInfo", storeInfo);
        return this;
    }
    /**
     * Provide DSL children to this orderStatus element.
     */
    withChildren(child) {
        this.addChild(child);
        return this;
    }
}
exports.OrderStatusUIBuilder = OrderStatusUIBuilder;
