// @dainprotocol/utils/src/generateUI/constants/uiAssets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UI_ASSETS = void 0;
exports.UI_ASSETS = {
    oauth: {
        google: {
            logo: 'https://www.google.com/calendar/images/calendar_logo_sm.gif',
            defaultTitle: 'Google Authentication',
            defaultContent: 'Please authenticate with Google to continue'
        },
        github: {
            logo: 'https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png',
            defaultTitle: 'GitHub Authentication',
            defaultContent: 'Please authenticate with GitHub to continue'
        },
        twitter: {
            logo: 'https://abs.twimg.com/responsive-web/client-web/icon-ios.b1fc727a.png',
            defaultTitle: 'Twitter Authentication',
            defaultContent: 'Please authenticate with Twitter to continue'
        },
        linkedin: {
            logo: 'https://static.licdn.com/sc/h/al2o9zrvru7aqj8e1x2rzsrca',
            defaultTitle: 'LinkedIn Authentication',
            defaultContent: 'Please authenticate with LinkedIn to continue'
        },
        telegram: {
            logo: 'https://telegram.org/img/t_logo.png',
            defaultTitle: 'Telegram Authentication',
            defaultContent: 'Please authenticate with Telegram to continue'
        }
    },
};
