// @dainprotocol/utils/src/generateUI/builders/server/oauth/oauth.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OAuthUIBuilder = void 0;
const base_builder_1 = require("../../../base/base.builder");
const uiAssets_1 = require("../../../constants/uiAssets");
function getProviderDefaults(provider) {
    // We'll do a quick lookup to see if we have assets for the given provider.
    // If not found, return an empty object or some fallback defaults.
    const providerAssets = uiAssets_1.UI_ASSETS.oauth[provider];
    if (!providerAssets) {
        return {};
    }
    return {
        logo: providerAssets.logo,
        title: providerAssets.defaultTitle,
        content: providerAssets.defaultContent,
    };
}
class OAuthUIBuilder extends base_builder_1.DSLBuilder {
    constructor(initialProps) {
        super("oauth2", initialProps);
        // If we have a known provider, apply defaults if that property is missing.
        if (initialProps === null || initialProps === void 0 ? void 0 : initialProps.provider) {
            const defaults = getProviderDefaults(initialProps.provider);
            if (!initialProps.logo && defaults.logo) {
                this.logo(defaults.logo);
            }
            if (!initialProps.title && defaults.title) {
                this.title(defaults.title);
            }
            if (!initialProps.content && defaults.content) {
                this.content(defaults.content);
            }
        }
    }
    /**
     * Sets the OAuth provider.
     */
    provider(provider) {
        return this.setProp("provider", provider);
    }
    /**
     * Sets the authentication URL.
     */
    url(url) {
        return this.setProp("url", url);
    }
    /**
     * Sets the title for the OAuth element.
     */
    title(title) {
        return this.setProp("title", title);
    }
    /**
     * Sets the content for the OAuth element.
     */
    content(content) {
        return this.setProp("content", content);
    }
    /**
     * Sets the logo for the OAuth element.
     */
    logo(logo) {
        return this.setProp("logo", logo);
    }
}
exports.OAuthUIBuilder = OAuthUIBuilder;
